node/src/components/consensus/era_supervisor.rs

```diff
@@ -193,16 +202,48 @@ where
             .iter_past(current_era, bonded_eras * 3)
             .collect();
 
-        info!(?era_ids, "collecting switch blocks");
+        // Asynchronously collect the information needed to initialize all recent eras.
+        let effects = async move {
+            info!(?era_ids, "collecting key blocks");
+            let key_blocks = effect_builder
+                .collect_key_blocks(era_ids.iter().cloned())
+                .await
+                .expect("should have all the key blocks in storage");
 
-        let effects = effect_builder
-            .collect_key_blocks(era_ids)
-            .event(move |key_blocks| Event::InitializeEras {
-                key_blocks: key_blocks.expect("should have all the switch blocks in storage"),
-                validators,
+            if current_era > activation_era_id + bonded_eras * 2 {
```

Comment:
Can we extract this to a method with a good name?

Reply:
It's inside an `async` block that can't have a reference to the whole `EraSupervisor` (unless we put everything into an `Arc<Mutex<_>>`), so it can't be a method.